### PR TITLE
LinghtningCLI now will not allow setting a class instance as a default

### LINKED ISSUE #18731: LightningCLI `trainer_defaults` get dumped as Python object
### Bug description

When providing an object like the logger to the `trainer_defaults`, it gets dumped as a Python object to the config file, which results in an invalid config file to load back into the CLI.

```
trainer:
  logger: <lightning.pytorch.loggers.wandb.WandbLogger object at 0x16e4db050>
```

### What version are you seeing the problem on?

v2.0, master

### How to reproduce the bug

```python
import torch
from lightning.pytorch import LightningModule
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch.cli import LightningCLI
from lightning.pytorch.loggers import WandbLogger


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def training_step(self, batch, batch_idx):
        return self.layer(batch).sum()

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)


if __name__ == "__main__":
    LightningCLI(BoringModel, trainer_defaults={
        "max_steps": 1,
        "logger": WandbLogger(project="my-project")
    })
```

### Error messages and logs

Inspect the config.yaml file in the current working directory. It contains the logger dumped as an object. It's unclear what the resolution to this could be. Ideally we would want

```
logger:
    - class_path: lightning.pytorch.loggers.wandb.WandbLogger
       init_args:
           ....
```

in the config file, but since we can't know the init args that were provided at the time of instantiation, it's hard to make this work. For the time being, it might be better to just drop the entry with a warning so that loading the config doesn't fail. 

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0): 2.1.0dev
#- Lightning App Version (e.g., 0.5.2): -
#- PyTorch Version (e.g., 2.0): 2.1
#- Python version (e.g., 3.9): 3.11
#- OS (e.g., Linux): MacOS
#- CUDA/cuDNN version: -
#- GPU models and configuration: -
#- How you installed Lightning(`conda`, `pip`, source): source
#- Running environment of LightningApp (e.g. local, cloud): local
```

</details>

### More info

jsonargparse==4.25.0

cc @carmocca @mauvilsa

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

Fixes #18731

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18822.org.readthedocs.build/en/18822/

### CODE REVIEW

File: tests/tests_pytorch/test_cli.py
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1]: I have noticed that some tests fail. All tests were successful for me locally so not sure what is the difference. Anyway, I don't have more time now and will look at it later.

**Review by awaelchli — APPROVED** [review]
Thank you @mauvilsa, this is great!

File: tests/tests_pytorch/test_cli.py
Comment (mauvilsa) [reply tests/tests_pytorch/test_cli.py:1]: The failing tests have been fixed now.

**Review by carmocca — APPROVED** [review]
Thank you!
